src/core/annotation.js

```diff
@@ -920,20 +935,98 @@ class TextWidgetAnnotation extends WidgetAnnotation {
   }
 
-  getOperatorList(evaluator, task, renderForms) {
-    if (renderForms || this.appearance) {
-      return super.getOperatorList(evaluator, task, renderForms);
-    }
+  getOperatorList(evaluator, task, renderForms, annotationStorage) {
+    const self = this;
+    return this.getAppearance(evaluator, task, annotationStorage).then(
+      content => {
+        if (renderForms || (self.appearance && content === "")) {
+          return super.getOperatorList(
+            evaluator,
+            task,
+            renderForms,
+            annotationStorage
+          );
+        }
 
-    const operatorList = new OperatorList();
+        const operatorList = new OperatorList();
+
+        // Even if there is an appearance stream, ignore it. Self is the
+        // behaviour used by Adobe Reader.
+        if (!self.data.defaultAppearance || content === "") {
+          return Promise.resolve(operatorList);
+        }
+
+        const matrix = [1, 0, 0, 1, 0, 0];
+        const bbox = [
+          0,
+          0,
+          self.data.rect[2] - self.data.rect[0],
+          self.data.rect[3] - self.data.rect[1],
+        ];
+
+        const transform = getTransformMatrix(self.data.rect, bbox, matrix);
+        operatorList.addOp(OPS.beginAnnotation, [
+          self.data.rect,
+          transform,
+          matrix,
+        ]);
+
+        const stream = new Stream(stringToBytes(content));
+        return evaluator
+          .getOperatorList({
+            stream,
+            task,
+            resources: self.fieldResources,
+            operatorList,
+          })
+          .then(function () {
+            operatorList.addOp(OPS.endAnnotation, []);
+            return operatorList;
+          });
+      }
+    );
+  }
+
+  getAppearance(evaluator, task, annotationStorage) {
+    // TODO: handle password: we mustn't write it in clear
+    if (!annotationStorage) {
+      return Promise.resolve("");
+    }
+    const value = annotationStorage[this.data.id] || "";
+    if (value === "") {
+      return Promise.resolve("");
+    }
+    const x = 2;
+    const y = 2;
+    if (this.data.comb) {
+      const width = this.data.rect[2] - this.data.rect[0];
+      const combWidth = (width / this.data.maxLen).toFixed(2);
+      let buf = `/Tx BMC q BT ${this.data.defaultAppearance} 1 0 0 1 ${x} ${y} Tm`;
+      let first = true;
+      for (const c of value) {
+        if (first) {
+          buf += ` (${c}) Tj`;
+          first = false;
+        } else {
+          buf += ` ${combWidth} 0 Td (${c}) Tj`;
+        }
+      }
+      buf += " ET Q EMC";
+      return Promise.resolve(buf);
+    }
 
-    // Even if there is an appearance stream, ignore it. This is the
-    // behaviour used by Adobe Reader.
-    if (!this.data.defaultAppearance) {
-      return Promise.resolve(operatorList);
+    const defaultAppearance = this.data.defaultAppearance;
+    const app = `/Tx BMC q BT ${defaultAppearance} 1 0 0 1 ${x} ${y} Tm (${value}) Tj ET Q EMC`;
```

Comment:
Let's call this `appearance` because `app` may indicate a different meaning. If the line becomes too long, it can be split over multiple lines with the template literals.